File: codes/ruby/polymorphism/overloading.rb
#!/usr/bin/ruby -w

class Person
    def read(text=nil)
        if text == nil then
            puts "I am reading"
        else
            if text.is_a?(String) then
                puts "A text: #{text}"
            end
        end
    end
end

class Student < Person
    def read(*args)
        case args.length
        when 0
            super
        when 1
            super(args[0])
        when 2
            puts "I am reading: " + args[0].to_s + " for " + args[1].to_s + " times"
        end
    end
end

puts "PERSON"
pe = Person.new
pe.read
pe.read("I am a person")

puts "STUDENT"
st = Student.new
st.read
st.read("I am a person")
st.read("a book", 5)

File: codes/python/polymorphism/overloading.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

class Person(object):
    def read(self, text=None):
        if text == None:
            print("I am reading")
        elif type(text) is str:
            print("A text: " + text)

class Student(Person):
    def read(self, *args):
        if len(args) == 0:
            Person.read(self)
        elif len(args) == 1:
            Person.read(self, args[0])
        else:
            print("I am reading: " + str(args[0]) + " for " + str(args[1]) + " times")

print("PERSON")
pe = Person()
pe.read()
pe.read("I am a person")

print("STUDENT")
st = Student()
st.read()
st.read("I am a student")
st.read("a book", 5)
